File: ui/bootstrap.py

import sys
import logging
import time
import traceback
import io
from pathlib import Path
from config_system import config

# Logging - output to BOTH file and console
def setup_logging():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s [%(name)s] %(levelname)s: %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S',
        handlers=[
            logging.FileHandler('nebula_debug.log', mode='w', encoding='utf-8'),
            logging.StreamHandler(sys.stdout)  # Also print to console
        ]
    )
    return logging.getLogger("ZenAI")

# --- Global Crash Handler ---
def handle_exception(exc_type, exc_value, exc_traceback):
    if issubclass(exc_type, KeyboardInterrupt):
        sys.__excepthook__(exc_type, exc_value, exc_traceback)
        return
    try:
        with open("ui_fatal_crash.txt", "w", encoding='utf-8') as f:
            f.write(f"Timestamp: {time.ctime()}\n")
            traceback.print_exception(exc_type, exc_value, exc_traceback, file=f)
    except (OSError, IOError):
        pass  # Can't write crash log, silently continue

def setup_crash_handler():
    sys.excepthook = handle_exception

def initialize_services():
    """Initialize all backend services (RAG, Memory, Features, Cleanup)."""
    logger = logging.getLogger("ZenAI.Bootstrap")
    
    # 1. Config Mode
    ZENA_MODE = config.zena_mode_enabled
    ZENA_CONFIG = {
        'enabled': config.zena_mode_enabled,
        'rag_enabled': True,
        'rag_source': 'knowledge_base',
        'swarm_enabled': config.swarm_enabled
    }
    logger.info(f"[Core] v3.1 Spec Initialization | Mode: {'Native' if ZENA_MODE else 'Legacy'} | Swarm: {config.swarm_enabled}")

    # 2. RAG System
    rag_system = None
    universal_extractor = None
    if ZENA_MODE:
        try:
            from zena_mode import LocalRAG
            from zena_mode.universal_extractor import UniversalExtractor
            
            # Universal Extractor for PDF/Image OCR in chat
            universal_extractor = UniversalExtractor()
            logger.info("[RAG] Universal Extractor ready for OCR")

            rag_cache = config.BASE_DIR / "rag_cache"
            rag_cache.mkdir(exist_ok=True)

            # RAG system initialization
            rag_system = LocalRAG(cache_dir=rag_cache)
            # Default RAG to enabled in config if not present
            if 'rag_enabled' not in ZENA_CONFIG:
                ZENA_CONFIG['rag_enabled'] = True
            logger.info("[RAG] RAG system initialized")
        except Exception as e:
            logger.error(f"[RAG] Failed to initialize: {e}")

    # 3. Conversation Memory
    conversation_memory = None
    try:
        from zena_mode import ConversationMemory
        conv_cache = config.BASE_DIR / "conversation_cache"
        conv_cache.mkdir(exist_ok=True)
        
        conversation_memory = ConversationMemory(cache_dir=conv_cache)
        logger.info("[Memory] Conversation memory initialized")
    except Exception as e:
        logger.error(f"[Memory] Failed to initialize conversation memory: {e}")

    # 4. Feature Detection
    from feature_detection import get_feature_detector
    feature_detector = get_feature_detector()
    logger.info("[Features] Feature detection complete")

    # 5. Cleanup Policy
    from cleanup_policy import get_cleanup_policy
    upload_cleanup = get_cleanup_policy(config.BASE_DIR / "uploads")
    logger.info("[Cleanup] Upload cleanup policy initialized")
    
    # 6. Help System
    try:
        if rag_system:
            from zena_mode.help_system import index_internal_docs
            from zena_mode.help_system import index_internal_docs
            # index_internal_docs(config.BASE_DIR, rag_system) # Blocking startup, disabled for debugging
            logger.info("[HelpSystem] Internal documentation indexing skipped (DEBUG)")
    except Exception as e:
        logger.warning(f"[HelpSystem] Failed to index internal docs: {e}")

    return {
        "rag_system": rag_system,
        "universal_extractor": universal_extractor,
        "conversation_memory": conversation_memory,
        "feature_detector": feature_detector,
        "upload_cleanup": upload_cleanup,
        "ZENA_CONFIG": ZENA_CONFIG,
        "ZENA_MODE": ZENA_MODE
    }

File: rust_output/src/bootstrap.rs
use anyhow::{Result, Context};
use crate::config_system::{config};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};

/// Setup logging.
pub fn setup_logging() -> () {
    // Setup logging.
    logging::basicConfig(/* level= */ logging::INFO, /* format= */ "%(asctime)s [%(name)s] %(levelname)s: %(message)s".to_string(), /* datefmt= */ "%Y-%m-%d %H:%M:%S".to_string(), /* handlers= */ vec![logging::FileHandler("nebula_debug.log".to_string(), /* mode= */ "w".to_string(), /* encoding= */ "utf-8".to_string()), logging::StreamHandler(sys::stdout)]);
    logging::getLogger("ZenAI".to_string())
}

/// Handle exception.
pub fn handle_exception(exc_type: String, exc_value: String, exc_traceback: String) -> Result<()> {
    // Handle exception.
    if issubclass(exc_type, KeyboardInterrupt) {
        sys::__excepthook__(exc_type, exc_value, exc_traceback);
        return;
    }
    // try:
    {
        let mut f = File::create("ui_fatal_crash.txt".to_string())?;
        {
            f.write(format!("Timestamp: {}\n", time::ctime()));
            traceback.print_exception(exc_type, exc_value, exc_traceback);
        }
    }
    // except (OSError, IOError) as _e:
}

pub fn setup_crash_handler() -> () {
    sys::excepthook = handle_exception;
}

/// Helper: setup phase for initialize_services.
pub fn _do_initialize_services_setup() -> Result<()> {
    // Helper: setup phase for initialize_services.
    let mut logger = logging::getLogger("ZenAI.Bootstrap".to_string());
    let mut ZENA_MODE = config::zena_mode_enabled;
    let mut ZENA_CONFIG = HashMap::from([("enabled".to_string(), config::zena_mode_enabled), ("rag_enabled".to_string(), true), ("rag_source".to_string(), "knowledge_base".to_string()), ("swarm_enabled".to_string(), config::swarm_enabled)]);
    logger.info(format!("[Core] v3.1 Spec Initialization | Mode: {} | Swarm: {}", if ZENA_MODE { "Native".to_string() } else { "Legacy".to_string() }, config::swarm_enabled));
    let mut rag_system = None;
    let mut universal_extractor = None;
    if ZENA_MODE {
        // try:
        {
            // TODO: from zena_mode import LocalRAG
            // TODO: from zena_mode.universal_extractor import UniversalExtractor
            let mut universal_extractor = UniversalExtractor();
            logger.info("[RAG] Universal Extractor ready for OCR".to_string());
            let mut rag_cache = (config::BASE_DIR / "rag_cache".to_string());
            rag_cache.create_dir_all();
            let mut rag_system = LocalRAG(/* cache_dir= */ rag_cache);
            if !ZENA_CONFIG.contains(&"rag_enabled".to_string()) {
                ZENA_CONFIG["rag_enabled".to_string()] = true;
            }
            logger.info("[RAG] RAG system initialized".to_string());
        }
        // except Exception as e:
    }
    Ok((ZENA_CONFIG, ZENA_MODE, logger, rag_system, universal_extractor))
}

/// Initialize all backend services (RAG, Memory, Features, Cleanup).
pub fn initialize_services() -> Result<()> {
    // Initialize all backend services (RAG, Memory, Features, Cleanup).
    let (mut ZENA_CONFIG, mut ZENA_MODE, mut logger, mut rag_system, mut universal_extractor) = _do_initialize_services_setup();
    let mut conversation_memory = None;
    // try:
    {
        // TODO: from zena_mode import ConversationMemory
        let mut conv_cache = (config::BASE_DIR / "conversation_cache".to_string());
        conv_cache.create_dir_all();
        let mut conversation_memory = ConversationMemory(/* cache_dir= */ conv_cache);
        logger.info("[Memory] Conversation memory initialized".to_string());
    }
    // except Exception as e:
    // TODO: from feature_detection import get_feature_detector
    let mut feature_detector = get_feature_detector();
    logger.info("[Features] Feature detection complete".to_string());
    // TODO: from cleanup_policy import get_cleanup_policy
    let mut upload_cleanup = get_cleanup_policy((config::BASE_DIR / "uploads".to_string()));
    logger.info("[Cleanup] Upload cleanup policy initialized".to_string());
    // try:
    {
        if rag_system {
            // TODO: from zena_mode.help_system import index_internal_docs
            // TODO: from zena_mode.help_system import index_internal_docs
            logger.info("[HelpSystem] Internal documentation indexing skipped (DEBUG)".to_string());
        }
    }
    // except Exception as e:
    Ok(HashMap::from([("rag_system".to_string(), rag_system), ("universal_extractor".to_string(), universal_extractor), ("conversation_memory".to_string(), conversation_memory), ("feature_detector".to_string(), feature_detector), ("upload_cleanup".to_string(), upload_cleanup), ("ZENA_CONFIG".to_string(), ZENA_CONFIG), ("ZENA_MODE".to_string(), ZENA_MODE)]))
}
